server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,345 @@
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Ensures that any writes to these files are moved to stable storage (made durable).
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void sync(Collection<String> names) throws IOException {
+        logger.trace("sync() called {}", names);
+        writeLock.lock();
+        try {
+            Collection<String> newLocalFiles = new ArrayList<>();
+            Collection<String> remoteFiles = Arrays.asList(getRemoteFiles());
+            for (String name : names) {
+                if (remoteFiles.contains(name) == false) newLocalFiles.add(name);
+            }
+            logger.trace("Synced files : {}", newLocalFiles);
+            localDirectory.sync(newLocalFiles);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Ensures that directory metadata, such as recent file renames, are moved to stable storage.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void syncMetaData() throws IOException {
+        logger.trace("syncMetaData() called ");
+        writeLock.lock();
+        try {
+            localDirectory.syncMetaData();
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Renames {@code source} file to {@code dest} file where {@code dest} must not already exist in
+     * the directory.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void rename(String source, String dest) throws IOException {
+        logger.trace("rename() called {}, {}", source, dest);
+        writeLock.lock();
+        try {
+            localDirectory.rename(source, dest);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Opens a stream for reading an existing file.
+     * Check whether the file is present locally or in remote and return the IndexInput accordingly
+     * @param name the name of an existing file.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public IndexInput openInput(String name, IOContext context) throws IOException {
+        logger.trace("openInput() called {}", name);
+        writeLock.lock();
+        try {
+            if (Arrays.asList(getRemoteFiles()).contains(name) == false) {
+                // If file has not yet been uploaded to Remote Store, fetch it from the local directory
+                logger.trace("File found in disk");
+                return localDirectory.openInput(name, context);
+            } else {
+                // If file has been uploaded to the Remote Store, fetch it from the Remote Store in blocks via
+                // OnDemandCompositeBlockIndexInput
+                logger.trace("File to be fetched from Remote");
+                return new OnDemandCompositeBlockIndexInput(remoteDirectory, name, localDirectory, fileCache, context);
```

Comment:
Played with this a bit.  I think we can still use `OnDemandBlockSnapshotIndexInput`directly here:

```java
                logger.trace("File to be fetched from Remote");
                RemoteSegmentMetadata remoteSegmentMetadata = remoteDirectory.readLatestMetadataFile();
                RemoteSegmentStoreDirectory.UploadedSegmentMetadata uploadedSegmentMetadata = remoteSegmentMetadata.getMetadata().get(name);
                BlobStoreIndexShardSnapshot.FileInfo fileInfo = new BlobStoreIndexShardSnapshot.FileInfo(
                    name,
                    new StoreFileMetadata(name, uploadedSegmentMetadata.getLength(),
                        uploadedSegmentMetadata.getChecksum(), null),
                    new ByteSizeValue(Long.MAX_VALUE)
                );
                return new OnDemandBlockSnapshotIndexInput(fileInfo, localDirectory, transferManager);
```

We would need to refactor TransferManager to use a functional interface instead of BlobContainer directly - something like:

```java
    @FunctionalInterface
    public interface BlobStreamReader {
       InputStream read(String name, long position, long length) throws IOException;
    }
    public TransferManager(final BlobStreamReader reader, final FileCache fileCache) {
        this.blobContainer = reader;
        this.fileCache = fileCache;
    }
```

and in CompositeDirectory's ctor:
```java
        this.transferManager = new TransferManager((name, position, length) -> new InputStreamIndexInput(remoteDirectory.openInput(name, new BlockedIOContext(position, length)), length), fileCache);
```

This would be nice for reuse and still adheres remoteSegmentStoreDirectory to the Directory interface. 
 Further if we wanted to fetch a file from a different remote repository we could.

Reply:
While we can explore this option, personally this seems like a very hacky way to me.

Also changing the current constructor of TransferManager will require changing the current RemoteSnapshot flow as well which we don't want to touch in this PR (without the unit and IT tests it has already touched 20 files with more than 1000 added/changed lines)

But agree on the reusability one, for the parts common to both RemoteSnapshot and CompositeDirectory, we will need to have common classes and will need to refactor them as well. (Can create another issue for this and discuss in more details there if possible)

For now I think the AbstractTransferManager as mentioned [here](https://github.com/opensearch-project/OpenSearch/pull/12782#discussion_r1569118039) seems the best way without modifying the constructor of current TransferManager. Keen to know your thoughts on this @mch2  ?

Reply:
> While we can explore this option, personally this seems like a very hacky way to me.

I'm curious why?  I think you'd ultimately have less to test here and reuse more code. You won't need OnDemandCompositeBlockIndexInput or a separate TransferManager impl.  `CompositeDirectoryTransferManager` from what I can tell logically does the same things as `TransferManager` in writing fetching the IndexInput and writing to an output stream. The only diff is the source of the input.

> Also changing the current constructor of TransferManager will require changing the current RemoteSnapshot flow as well which we don't want to touch in this PR (without the unit and IT tests it has already touched 20 files with more than 1000 added/changed lines)

Fair enough, but I don't think we should be scared to update an internal component to make it more generic.  It will also reduce lines added.

Reply:
Updated the Transfer Manager implementation to include a functional interface